### PR TITLE
Use go-connlimit to ratelimit with 429 responses

### PR DESCRIPTION
**Pull request by shoenig**
Expands on https://github.com/hashicorp/nomad/pull/9359 from @benbuzbee 

In addition this PR fixes the original tests, which produced the correct result for the wrong reason. In TLS mode, the clients were creating HTTP connections which would be shunted by the server in the same way being ratelimited would (i.e. `io.EOF`). Instead, use a tls `net.Conn` for tls tests.

Also rebased the original commit to remove extraneous file. 

Closes #9359

### CODE REVIEW

File: command/agent/http_test.go [hunk 1]
```diff
@@ -1036,18 +1049,37 @@ func TestHTTPServer_Limits_OK(t *testing.T) {
...
+
+	dial := func(addr string, useTLS bool) net.Conn {
+		if useTLS {
+			cert, err := tls.LoadX509KeyPair(foocert, fookey)
+			require.NoError(t, err)
+			conn, err := tls.Dial("tcp", addr, &tls.Config{
+				Certificates:       []tls.Certificate{cert},
+				InsecureSkipVerify: true, // good enough
+			})
+			require.NoError(t, err)
```
Comment (tgross) [inline command/agent/http_test.go:1066]: If we don't pass the `testing.T` from the test case into this function and it's closing over the top-level `testing.T`, I think we lose the association with the test case and it makes errors harder to track down. Might want to either pass in the test case's `testing.T` or have the function return the error and do the assertion in the caller.

**Review by tgross — APPROVED** [review]
LGTM!

File: command/agent/http_test.go (hunk 1 above)
Comment (shoenig) [reply command/agent/http_test.go:1066]: Good catch, fixed!

**Review by schmichael — COMMENTED** [review]
Yikes, great catch in the tests.

I'm still concerned about this approach for 2 reasons:

1. Writing a response before reading the request will not work with every HTTP client.
2. Allowing an attacker to block new connections _from other addresses_ seems like a DoS. Even if it's only 10ms I don't think there's anything preventing an attacker from filling the kernel's listen backlog with pending connections each of which will cause a 10ms timeout.

I think 1. will make 2 worse. Imagine a Python script which uses a synchronous http library that only reads after successfully writing. Perhaps this script is run as a system job across all nodes to check a bit of Nomad state and react appropriately. This works fine until the user deploys this system in a remote region behind a NAT. The NAT makes every incoming connection from the same IP, so the limit is reached. However the synchronous Python http library hits the 10ms timeout every time because it will never read until its write succeeds (although the write "succeeding" probably depends on its size and various buffers between the systems... so the simple case of a GET may work as expected?).

If there are 500 remote nodes, they could easily DoS the cluster by filling the accept backlog with 500 10ms timeouts. Operators may be unable to use Log Monitoring (either via the UI or API) becaue of the unintnetional DoS. (#7893 would partially solve this.)

There's a lot of socket/TCP subtletly involved here that maybe I'm wrong about? Maybe this is safe? Confirming via testing sounds incredibly difficult considering the wide variety of http clients, requests, socket implementations, and middleboxes involved.

Regardless I think a better solution from a safety **and** usability standpoint would be to implicitly create soft and hard limits:

The soft limit would read the request and respond with a 429 + normal timeout.

The hard limit would behave as before: immediately reject the connection.

The hard limit could implicitly be set to the soft limit + 1. I can't imagine 1 extra connection over the limit will break anybody's security model.

This would allow us to keep all of the existing hard limit code intact, and add the soft limit as a plain old HTTP handler.

**Comment by tgross** [comment]
That reasoning all makes sense to me; we can't assume clients are well-behaved and non-blocking.

I like the idea of keeping the hard limit but adding a soft limit handler. But I might adjust the relationship of the hard limit and soft limit. In the accidental case you described with 500 nodes, only a single one of those in flight requests would get a HTTP429 before the remaining connections get dropped. And if the client is blocking as you've described, no one will see the HTTP429 before we start dropping connections. So maybe we want to give ourselves a little more breathing room than soft limit + 1? (Like 10% or something?)

**Comment by schmichael** [comment]
Ah good point. `hard_limit = soft_limit + 1` only really helps if it's a single client hitting the limit. 10% seems good to me. Linux+Go are quite good at handling a large number of connections, and our default of 100 is pretty low, so I don't foresee an extra 10 being a big deal.

**Pull request merged**